src/consul.rs

```diff
@@ -0,0 +1,308 @@
+    /// This function retrieves the state of all DNS records stored under a specific Consul key.
+    pub async fn fetch_all_dns_records(&self) -> Result<HashMap<String, DnsRecord>> {
+        let url = self.kv_api_base_url.join("dns_records")?;
+        let resp = self.http_client.get(url).send().await?;
+
+        let mut records: HashMap<String, DnsRecord> = HashMap::new();
+        if !resp.status().is_success() {
+            eprintln!("Failed to fetch DNS records: {}", resp.status());
+            if resp.status() == StatusCode::NOT_FOUND {
+                return Ok(records);
+            }
+            return Err(anyhow::anyhow!(
+                "Failed to fetch DNS records: {}",
+                resp.status()
+            ));
+        }
+
+        let body = resp.bytes().await?;
+
+        let kv_response: Vec<ConsulKVResponse> = serde_json::from_slice(&body)
+            .map_err(|e| anyhow::anyhow!("Failed to decode KV response: {}", e))?;
+
+        for entry in kv_response {
+            if let Some(encoded_value) = entry.value {
+                let decoded_bytes = &BASE64_STANDARD
+                    .decode(encoded_value)
+                    .expect("Can't decode base64");
+
+                // Deserialize the JSON string to a HashMap<String, DnsRecord>
+                let record_map: HashMap<String, DnsRecord> = serde_json::from_slice(decoded_bytes)
+                    .map_err(|e| anyhow::anyhow!("Failed to deserialize DnsRecord: {}", e))?;
+
+                records.extend(record_map);
+            }
+        }
+
+        Ok(records)
+    }
+}
+
+fn parse_dns_tags(tags: Vec<String>) -> Vec<DnsRecord> {
+    const PREFIX: &str = "external-dns.";
+    // Parse service tags of the format `external-dns.<field>=<value>`.
+    let mut dns_tags: HashMap<String, HashMap<String, String>> = HashMap::new();
+    for tag in tags.into_iter() {
+        let Some(rest) = tag.strip_prefix(PREFIX) else {
+            continue;
+        };
+        let Some((identifier, rest)) = rest.split_once('.') else {
+            continue;
+        };
+        let Some((field, value)) = rest.split_once('=') else {
+            continue;
+        };
+        dns_tags
+            .entry(identifier.to_string())
+            .or_default()
+            .insert(field.to_string(), value.to_string());
+    }
+
+    let mut records = Vec::new();
+    for (identifier, mut tags) in dns_tags {
+        let Some(hostname) = tags.remove("hostname") else {
+            println!("Missing hostname for identifier: {}", identifier);
+            continue;
+        };
+        let type_: DnsType = match tags.remove("type").map(|t| t.parse()) {
+            None => {
+                println!("Missing type for identifier: {}", identifier);
+                continue;
+            }
+            Some(Ok(t)) => t,
+            Some(Err(e)) => {
+                eprintln!("Failed to parse type for identifier {}: {}", identifier, e);
```

Comment:
Something like "Unsupported record type {type} specified for identifier {identifier}" would be a more user-friendly error message.